controllers/nim_account_controller.go

```diff
@@ -0,0 +1,424 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/go-logr/logr"
+	"github.com/kuadrant/authorino/pkg/log"
+	"github.com/opendatahub-io/odh-model-controller/api/nim/v1"
+	"github.com/opendatahub-io/odh-model-controller/controllers/constants"
+	"github.com/opendatahub-io/odh-model-controller/controllers/utils"
+	templatev1 "github.com/openshift/api/template/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/reference"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type (
+	NimAccountReconciler struct {
+		client.Client
+		Log logr.Logger
+	}
+)
+
+const (
+	apiKeySpecPath = "spec.apiKeySecret.name"
+)
+
+func (r *NimAccountReconciler) SetupWithManager(mgr ctrl.Manager, ctx context.Context) error {
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &v1.Account{}, apiKeySpecPath, func(obj client.Object) []string {
+		return []string{obj.(*v1.Account).Spec.APIKeySecret.Name}
+	}); err != nil {
+		r.Log.Error(err, "failed to set cache index")
+		return err
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("odh-nim-controller").
+		For(&v1.Account{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.Secret{}).
+		Owns(&templatev1.Template{}).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(
+			func(ctx context.Context, obj client.Object) []reconcile.Request {
+				var requests []reconcile.Request
+				accounts := &v1.AccountList{}
+				if err := mgr.GetClient().List(ctx, accounts, client.MatchingFields{apiKeySpecPath: obj.GetName()}); err != nil {
+					r.Log.Error(err, "failed to fetch accounts")
+					return requests
+				}
+				for _, item := range accounts.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+						Name:      item.Name,
+						Namespace: item.Namespace,
+					}})
+				}
+				return requests
+			})).
+		Complete(r)
+}
+
+func (r *NimAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("Account", req.Name, "namespace", req.Namespace)
+	ctx = log.IntoContext(ctx, logger)
+
+	account := &v1.Account{}
+	if err := r.Client.Get(ctx, req.NamespacedName, account); err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.V(1).Info("account deleted")
+			r.cleanupResources(ctx, req.Namespace)
```

Comment:
it seems that the resources affected by the cleanup have the `Account` as owner, I guess they get cleaned up automatically by k8s gc, so may not be needed ?

Reply:
Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/7929b53bb3cbda70594b2aa0169f26d34bf92722.

Reply:
You know what @lburgazzoli, I'm just reliazing something Yes,  you were right, we didn't need to delete the resources when an acount is being deleted, k8s will handle this.
But the func I removed in the above commit, was also used for when the API key secret is deleted. And this still needs to be handled.

So, I need to bring back the cleanup utility func  for cleanups, but only use it when the secret is deleted.

Reply:
(re)Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/15dccb2c6f0d1b9814b7a5fe1ba75c8c7a6d4723.